src/dinitctl.cc

```diff
@@ -1903,39 +1903,61 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, service_dir_
     }
 
     if (socknum >= 0) {
+        if (verbose) {
+            cout << "Service '" << to << "' has been " << (enable ? "enabled" : "disabled") << "." << endl;
+        }
+
         // Check status of the service now
         auto m = membuf()
                 .append((char)cp_cmd::SERVICESTATUS)
                 .append(to_handle);
         write_all_x(socknum, m);
 
-        wait_for_reply(rbuffer, socknum);
-        if (rbuffer[0] != (char)cp_rply::SERVICESTATUS) {
+        int r = rbuffer.fill_to(socknum, 2);
+        while (r > 0 && rbuffer[0] >= 100) {
+            unsigned pktlen = (unsigned char) rbuffer[1];
+            fill_buffer_to(rbuffer, socknum, pktlen);
+
+            if (rbuffer[0] == (char)cp_info::SERVICEEVENT) {
+                wait_service_state(socknum, rbuffer, to_handle, to, !enable, verbose);
+                return 0;
+            }
+
+            rbuffer.consume(pktlen);
+            r = rbuffer.fill_to(socknum, 2);
+        }
+
+        if (r == -1) {
+            throw cp_read_exception(errno);
+        }
+        else if (r == 0) {
             cerr << "dinitctl: protocol error." << endl;
             return 1;
         }
-        rbuffer.consume(1);
+        else if (rbuffer[0] == (char)cp_rply::SERVICESTATUS) {
```

Comment:
You don't need `else` clause (here or on line 1933) because if the condition is true the function exits anyway (via `return` or `throw`).

It would be good to error out early, i.e. change 

    if (rbuffer[0] == (char)cp_rply::SERVICESTATUS) {

To

    if (rbuffer[0] != (char)cp_rply::SERVICESTATUS) {

... and handle the error, then leave the other code (normal execution) outside the `if`. (That has the nice side effect that you don't end up changing the indentation of a whole bunch of code, so the diff is smaller, though that's not the primary concern).